### PR TITLE
Bucket: expose Available() and Capacity()

### PR DESCRIPTION
**Pull request by hongchaodeng**
Hi,

The current token bucket implementation is great. Nonetheless we need some underlying data to implement more powerful functionalities.

For example, if we want to implement a function `CanAccept` which only checks available tokens without taking it; implement a function that exports rate limiter's metrics to better understand current flow and bottleneck on our services. Both functions will benefit if we can get the underlying data like `avail` and `capacity`, including the already public `Rate()`.

Let me know if there are questions. Highly appreciate it if anyone can help review this issue. Thanks!

### CODE REVIEW

File: ratelimit.go [hunk 1]
```diff
@@ -171,6 +171,16 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 	return count
 }
 
+// Available returns the number of available tokens
+func (tb *Bucket) Available() int64 {
+	return tb.avail
```
Comment (axw) [inline ratelimit.go]: tb.mu should be held while accessing tb.avail

**Comment by axw** [comment]
Seems fine to me, with one small fix; deferring to @rogpeppe, the author, though.

File: ratelimit.go [hunk 2]
```diff
@@ -171,6 +171,18 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 	return count
 }
 
+// Available returns the number of available tokens
+func (tb *Bucket) Available() int64 {
```
Comment (xiang90) [inline ratelimit.go]: I guess we'd better also doc that `It will be negative when there are consumers waiting for tokens.`

File: ratelimit.go [hunk 3]
```diff
@@ -171,6 +171,19 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 	return count
 }
 
+// Available returns the number of available tokens.
+// It will be negative when there are consumers waiting for tokens.
+func (tb *Bucket) Available() int64 {
+	tb.mu.Lock()
+	defer tb.mu.Unlock()
+	return tb.avail
```
Comment (rogpeppe) [inline ratelimit.go]: This is not right. The number of available tokens changes as time progresses
but the returned value from this function does not.

You need to adjust avail before returning the value. Something like
this:
	
	// Available returns the number of available tokens.
	// It will be negative when there are consumers waiting for tokens.
	// Note that this method is inherently racy - the returned
	// count may not be current a moment later.
	func (tb *Bucket) Available(now time.Time) int64 {
		return tb.available(time.Now())
	}
	
	// available is the internal version of available - it takes the current time as
	// an argument to enable easy testing.
	func (tb *Bucket) available(now time.Time) int64 {
		tb.mu.Lock()
		defer tb.mu.Unlock()
		tb.adjust(now)
		return tb.avail
	}

Also, please write a test (note the internal version above, as with the
other methods, written that way so it's easy to write a deterministic test).

File: ratelimit.go [hunk 4]
```diff
@@ -171,6 +171,19 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
...
+}
+
+// Capacity returns the capacity of this rate limiter.
```
Comment (rogpeppe) [inline ratelimit.go]: // Capacity returns the capacity that the bucket was created with.

**Comment by rogpeppe** [comment]
Thanks for this.
This is good in principle, but could do with a little work to make it correct and tested.

Please rebase to a single commit before re-proposing, thanks.

**Comment by hongchaodeng** [comment]
Thanks @rogpeppe for your reply. I realized my mistake and fixed it.
I have also added a test for the internal available().

I squash all I want to do in single commit. However, I add another commit to update `gocheck` and `math.Abs` -- the second one made the test not working.

Let me know if there are questions. Thanks again!

**Comment by hongchaodeng** [comment]
@rogpeppe 
Changed naming to `TestAvailable()`

**Comment by hongchaodeng** [comment]
@rogpeppe 
Addressed all comments. PTAL.

File: ratelimit_test.go [hunk 5]
```diff
@@ -125,6 +126,32 @@ var takeTests = []struct {
...
 
+var availTests = []struct {
+	capacity     int64
+	fillInterval time.Duration
+	take         int64
+	sleep        time.Duration
+
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{
```
Comment (rogpeppe) [inline ratelimit_test.go]: }{{

to save an indent level.

Also, please move the tests to just before TestAvailable.

File: ratelimit_test.go [hunk 6]
```diff
@@ -125,6 +126,32 @@ var takeTests = []struct {
...
+		fillInterval:          time.Second,
+		take:                  5,
+		sleep:                 time.Second,
+		expectCountAfterTake:  0,
+		expectCountAfterSleep: 1,
+	}, {
+		capacity:              2,
+		fillInterval:          time.Second,
+		take:                  1,
+		sleep:                 time.Second,
```
Comment (rogpeppe) [inline ratelimit_test.go]: I'd like to see at least one test that tests that available returns the correct thing when not called on a fillInterval boundary.

**Comment by rogpeppe** [comment]
Thanks very much for making these changes. LGTM modulo some comments above.

File: ratelimit_test.go [hunk 7]
```diff
@@ -125,6 +126,44 @@ var takeTests = []struct {
...
+	expectCountAfterTake:  0,
+	expectCountAfterSleep: 1,
+}, { // should fill plus existing count
+	capacity:              2,
+	fillInterval:          time.Second,
+	take:                  1,
+	sleep:                 time.Second,
+	expectCountAfterTake:  1,
+	expectCountAfterSleep: 2,
+}, { // shouldn't fill before interval
```
Comment (hongchaodeng) [inline ratelimit_test.go]: @rogpeppe 
Add two more tests to check boundaries. PTAL.

File: ratelimit_test.go [hunk 8]
```diff
@@ -125,6 +126,44 @@ var takeTests = []struct {
...
 
+var availTests = []struct {
+	capacity     int64
+	fillInterval time.Duration
+	take         int64
+	sleep        time.Duration
+
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{{ // should fill
```
Comment (rogpeppe) [inline ratelimit_test.go]: let's make this comment into an "about" field for consistency with
other table-driven tests, and log it. (it means that when a test
does fail, we can quickly find out which one has failed)

**Comment by rogpeppe** [comment]
LGTM with one trivial suggestion.

File: ratelimit_test.go [hunk 9]
```diff
@@ -125,6 +126,49 @@ var takeTests = []struct {
...
+	about        string
+	capacity     int64
+	fillInterval time.Duration
+	take         int64
+	sleep        time.Duration
+
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{{
+	about:                 "should fill tokens after interval",
```
Comment (hongchaodeng) [inline ratelimit_test.go]: @rogpeppe 
Add `about` in table and test failure output.

File: ratelimit.go [hunk 10]
```diff
@@ -171,6 +171,28 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 	return count
 }
 
+// Available returns the number of available tokens.
+// It will be negative when there are consumers waiting for tokens.
+// Note that this method is inherently racy - the returned count
```
Comment (liggitt) [inline ratelimit.go]: rather than "inherently racy" (since the available number is actually computed in a thread-safe way), maybe clarify that the returned number of available tokens does not guarantee calls like Take/TakeAvailable will succeed, and that this is primarily for introspection/metrics/reporting/debugging?

Comment (rogpeppe) [inline ratelimit.go]: That's a good idea, thanks.
@hongchaodeng How about this?

// Available returns the number of available tokens.
// It will be negative when there are consumers waiting for tokens.
// Note that if this returns greater than zero,
// it does not guarantee that calls that take tokens
// from the buffer will succeed, as the number of available
// tokens could have changed in the meantime.
// This method is intended primarily for metrics reporting
// and debugging.

Comment (rogpeppe) [inline ratelimit.go]: Actually, I'll land this now and make that change in another PR.

Comment (rogpeppe) [inline ratelimit.go]: Done.
